### DISCUSSION TITLE
[IfcOpenShell-Python] Are all shapes created with ifcopenshell.geom with settings.USE_PYTHON_OPENCASCADE an TopoDS_Compound?

### DISCUSSION THREAD

**Discussion by Martin15135215** [IfcOpenShell/IfcOpenShell#4845] [Q&A]
Here is a example script:
https://github.com/tpaviot/pythonocc-demos/blob/master/jupyter_notebooks/ifc_display_basic_file.ipynb

If i do add this in the script:
`print(pdct_shape.geometry.ShapeType)`

All is TopoDS_Compound:
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/110968398/0991a430-8a30-4d96-9280-d8d29256a225)

**Comment by aothms** [comment]
For create shape on a product: yes, because these correspond to the
individual items of the representation. You can also call create shape on
an individual representation item in that case you'll get something else.
Sent from a mobile device, excuse my brevity. Kind regards, Thomas

Op di 11 jun 2024 15:57 schreef Martin15135215 ***@***.***>:

**Reply by Martin15135215** [reply]
I see, so I changed the script from:
` pdct_shape = ifcopenshell.geom.create_shape(settings, inst=product)`

to:
```
pdct_shape = ifcopenshell.geom.create_shape(settings, inst=product.Representation.Representations[0])
print(pdct_shape)
```
Now, I got directly the shape, however it is still a TopoDS_Compound:
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/110968398/83dc0429-1ef5-4cac-8ce8-2507ae6611a5)

**Reply by Martin15135215** [reply]
I am wondering if i get a TopoDS_CompSolid or TopoDS_Solid instead of TopoDS_Compound

**Comment by Moult** [comment]
You're feeding it a `product.Representation.Representations[0]`. This is a shape representation, not a representation item. A shape representation can contain multiple items.

I'd also recommend using the `ifcopenshell.util.representation` module because hardcoding things like `.Representation.Representations[0]` can have issues (e.g. the 0th representation may not exist, or it may not be the body representation, or it may be a mapped representation, etc).

For example, you can do `rep = ifcopenshell.util.representation.get_representation(element, "Model", "Body", "MODEL_VIEW")` to specifically get the 3D body representation. Then you can do `items = ifcopenshell.util.representation.resolve_items(rep)` which will get you all of the items of that representation, successfully traversing things like mapped items.
